src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -38,6 +41,7 @@ private static void RootReflectionSerializerDependencies()
             {
                 s_defaultSimpleConverters = GetDefaultSimpleConverters();
                 s_defaultFactoryConverters = GetDefaultFactoryConverters();
+                s_converterRetrievalFunc = (type, options) => options.GetConverterForTypeReflection(type);
```

Comment:
cc @krwq 

Per our discussion yesterday, I think this PR is further evidence that we might want to abstract converter resolution behind contract resolvers. This would require either:

* Changing `ITypeInfoResolver` to an abstract class that contains a virtual method that resolves converters. The default resolver would then override that virtual doing reflection-based resolution.
* Hide the reflection-based contact resolution behind an internal method in `DefaultJsonTypeInfoResolver` and call that using a type test:

```C#
if (_resolver is DefaultJsonTypeInfoResolver defaultResolver)
{
     // calling into a reflection method with suppressed AOT warnings, 
     // the fact we have a reflection resolver stored in the options instance should
     // have triggered warnings elsewhere.
     return defaultResolver.GetConverterUsingReflection(typeToConvert, options);
}

return _resolver.GetTypeInfo(typeToConvert)?.PropertyInfoForTypeInfo?.ConverterBase;
```

Reply:
I agree they should be behind contract resolver but I do not agree that IJsonTypeInfo would have to become abstract. Currently source gen stores converters on JsonTypeInfos and we could do the same for reflection